RFC/src/RFC-0171_PeerToPeerMessagingProtocol.md

```diff
@@ -0,0 +1,662 @@
+- MUST sequentially try to connect to one the peer's [NetAddress]es until one succeeds or all fail using an `OutboundConnection`
+- MUST immediately reject and dispose of a multipart message not consisting of four parts, as detailed in [MessageEnvelope].
+- MUST construct a [MessageEnvelope] from the multiple parts.
+- MUST pass the constructed [MessageEnvelope] to the message handler.
+- Should a connection drop, the connection state MUST transition to `Connecting` and the connection retried.
+- When a shutdown signal is received, MUST send a `net::Disconnect` message and drop the connection.
+
+#### ConnectionManager
+
+The ConnectionManager manages a set of live PeerConnections and provides an abstraction for other components
+to initiate and use PeerConnections without having to worry about attaching the new PeerConnection to message handlers.
+
+It consists of a list of active peer connections and an `inproc` message handler socket. This socket is 'written to' whenever
+a message is received from any active [PeerConnection] for other components to act on.
+
+Methods may include:
+
+- `establish_connection(Peer)` - create and return a new PeerConnection
+- `disconnect(peer)` - disconnect a particular peer
+- `suspend()` - temporarily suspend connections
+- `resume()` - temporarily suspend connections
+- `shutdown` - cleanly shutdown all PeerConnections
+
+The `ConnectionManager`:
+
+- MUST call `suspend` on every [PeerConnection] if it's `suspend` method is called
+- MUST call `resume` on every [PeerConnection] if it's `resume` method is called
+- MUST call `shutdown` on every [PeerConnection] if it's `shutdown` method is called
+- MUST create a new [PeerConnection] with the given Peer and NetAddress, when `establish_connection` is called
+- MUST call `shutdown` on the [PeerConnection] and remove the connection for the given Peer, when `disconnect(peer)` is called
+- MAY disconnect peers if the connection has not been used for an extended period
+- SHOULD disconnect the least recently used peer if the connection pool greater than `max connections`
+
+#### ControlService
+
+The purpose of this service is to negotiate a new secure PeerConnection.
+
+The control service accepts a single message:
+- `peer::EstablishConnection(pk, curve_pk, net_address)`
+
+A ControlService:
+
+- MUST listen for connections on a predefined CONTROL PORT
+- SHOULD deny connections from banned peers
+
+To establish a peer connection, the following steps apply:
+
+Alice wants to connect to Bob
+
+1. Alice opens a new socket for the `PeerConnection` to be.
+    - A new CURVE keypair is generated
+2. Alice connects to Bob's control server and Bob accepts the connection.
+3. Alice sends an `peer::establish_connection` message, with
+    - the CURVE public key for the socket connection,
+    - the node's public key corresponding to its [Node ID],
+    - the [NetAddress] of the new PeerConnection.
```

Comment:
I hadn't thought of that,  and I agree - the target would just reject all the messages coming to it as they would be invalid for the control socket or for PeerConnections it would reject as message signatures would be invalid, but there is a DDoS vector . I think needs to be more to this connection negotiation.

Thoughts, off the top of my head:
- IP addresses would be easier to validate the source IP matches the new NetAddress, not so for onion/I2P but may help with overall security
- A (hopefully light-weight) request to the given NetAddress which verifies that whoever is on the other end has the private key that was given in the establish connection request. If that validation fails the attacker is immediately banned